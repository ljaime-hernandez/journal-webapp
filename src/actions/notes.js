import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";


export const startNewNote = () => {
    return async(dispatch, getState) => {

        // retrieves the user id from the store
        const uid = getState().auth.uid;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        // asynchronous function pushing a new note into the users journal
        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        // we both assign the new note as active on our notesReducer and set it
        // ready on the notesScreen component for us to input the information
        dispatch(activeNote(doc.id, newNote));
        dispatch(addNewNote(doc.id, newNote));
    }
}

// sets the note into an active state, used in the NoteScreen component
export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id, 
        ...note
    }
})

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id, 
        ...note
    }
})

// uses the loadNotes helper to retrieve an array of notes from the 
// users collection for it to be rendered on the sidebar
export const startLoadingNotes = (uid) => {
    return async(dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));

    } 
}

// sends an array of notes to the notesReducer for them to be loaded
export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async(dispatch, getState) => {

        // to save the note we retrieve the uid from the actual state on the
        // user store, as some users will forget or wont include a picture
        // on the note, then we make a small condition eliminating the
        // url value from the note
        const { uid } = getState().auth;

        if(!note.url){
            delete note.url;
        }

        // we declare the new note saving on a new variable, then proceeding for us to
        // delete the id, this is done so we reuse the previous note id to update it as
        // the note exists
        const noteToFirestore = {...note};
        delete noteToFirestore.id;

        // we make the request to the firebase db using the users uid, the note id and updating
        // all the items on its document with the update method
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        // the refresh note will trigger an update on all the existing notes so they are
        // arranged accordingly and they also display the appropriate information on the screen,
        // afterwards sending us a screen popup with Swal to confirm the note was succesfully saved
        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved', note.title, 'success');
    } 
}

export const refreshNote = (id, note) => ({

    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const fileUploading = (file) => {
    return async (dispatch, getState) => {
        
        // the fileUploading action can be called only on an active note, either it is
        // a new note or an existing one, so we use the getState method to retrieve
        // the active note information from the store
        const {active:activeNote} = getState().notes;

        // a quick pop up will show using Swal with a loading spinner, while the
        // file is being uploaded using the fileUpload helper which will return the URL
        // where the file is uploaded, followed we call the startSaveNote action which
        // will save the whole note to our firebase collection and finally the
        // pop up will close after the process is done
        Swal.fire({
            title: 'Uploading... please wait',
            text: 'Please wait...',
            allowOutsideClick: false,
            willOpen: () => {
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote))

        Swal.close();
    }
}

export const startDelete = (id) => {
    return async(dispatch, getState) => {

        // similar to our previous actions, we use the getState method to retrieve the
        // user uid, then we use both the note id and the user uid to access to its
        // collection in firebase and trigger the delete method, deleting the document
        // stored on it
        const uid =  getState().auth.uid;

        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const noteLogout = () => {
    
    return {
        type: types.notesLogoutClean
    }
}