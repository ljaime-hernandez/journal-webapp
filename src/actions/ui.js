import { types } from "../types/types";

export const setError = (error) => {
    return {
        type: types.uiSetError,
        payload: error
    }
}

export const removeError = () => {
    return {
        type: types.uiRemoveError,
    }
}

export const startLoading = () => {
    return {
        type :types.uiStartLoading
    }
}

export const finishLoading = () => {
    return {
        type :types.uiFinishLoading
    }
}