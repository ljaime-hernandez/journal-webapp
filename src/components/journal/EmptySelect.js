import React from 'react';

export const EmptySelect = () => {
  return (
    <div className="empty__main-content">
        <p>
            Select a Journal
            <br/>
            Create an entry
        </p>

        <i className="far fa-star fa-4x mt-5"></i>
    </div>
  )
};
