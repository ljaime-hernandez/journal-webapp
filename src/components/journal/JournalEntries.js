import React from 'react';
import { useSelector } from 'react-redux';
import { JournalEntry } from './JournalEntry';

export const JournalEntries = () => {

    // uses the information on the store for us to
    // render the existent notes in the users account on 
    // the sidebar, if not, the sidebar will still have a 
    // new entry button for the user to create it.
    const {notes} = useSelector( state => state.notes );

  return (
    <div className="journal__entries animate__animated animate__fadeInLeft">

        {
            notes.map(note => (
                <JournalEntry
                    key={note.id}
                    {...note}
                />
            ))
        }
    </div>
  )
};
