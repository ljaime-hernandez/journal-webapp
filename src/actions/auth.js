import { firebase, googleAuthProvider } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from './ui';
import { types } from "../types/types";
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// 
export const startLoginEmailPassword = (email, password) => {

    // the dispatch contains functions from different reducers, but the useDispatch hook
    // is linked to redux, using the environment developed in the store file along with
    // our redux functions, the reducers are contained in the combineReducer method, allowing
    // us to dispatch several actions at once
    return (dispatch) => {

        dispatch(startLoading());

        // account authentication from regular email accounts, it returns a callback which we will
        // use to dispatch the login action from our auth reducer and, either it goes thru or not
        // will also dispatch the finishLoading function to enable the LoginScreen button for login
        firebase.auth().signInWithEmailAndPassword(email, password)
        .then(({user}) => {
            dispatch(finishLoading());
            dispatch(login(user.uid, user.displayName));
        }).catch( e => {
            dispatch(finishLoading());
            Swal.fire('Error', e.message,'error');
        })
    }
}

export const startGoogleLogin = () => {

    // as this function is a callback, the dispatch will be sent back to our authReducer
    // environment for it to handle user login status
    return (dispatch) => {

        // the googleAuthProvider is an authentication method returned from our firebase
        // db, along with the signInWithPopup the user will be able to use its google
        // original account to create an account in our webpage, and for us to
        // use the information for authentication and also render a proper user interface
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                )
            })
        }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const logout = () => {
    return {
        type: types.logout
    }
}

export const startLogout = () => {
    return (dispatch) => {
        firebase.auth().signOut();
        dispatch(noteLogout());
        dispatch(logout());
    }
}
    
export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then( async ({user}) => {
            await user.updateProfile({displayName: name});

            dispatch(login(user.uid, user.displayName))
        }).catch( e => {
            console.log(e);
            // alert displaying an error message if theres errors on the user input
            // or if the email exists
            Swal.fire('Error', e.message,'error');
        })
    }
}

