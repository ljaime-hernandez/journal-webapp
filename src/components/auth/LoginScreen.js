import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

  const dispatch = useDispatch();
  const {ui} = useSelector(state => state);

  const {msgError, loading} = ui;


  const [values, handleInputChange] = useForm({
    email: 'email@email.com',
    password: '123456'
  })

  const {email, password} = values;

  const handleLogin = (e) => {
    e.preventDefault();

    if(isLoginValid()){
      dispatch(startLoginEmailPassword(email, password))
    }
  }

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  }

  const isLoginValid = () => {

    if ( !validator.isEmail(email)){
      dispatch(setError("The email input is incorrect"))
      return false;
    } else if( password < 5){
      dispatch(setError("Password should be longer than 4 characters"))
      return false;
    }

    dispatch(removeError())
    return true;
  }

  return (
  <>
      <h3 className="auth__title">Login</h3>

      <form 
        onSubmit={handleLogin}
        className='animate__animated animate__fadeIn'  
      >

      {
        msgError &&
        (<div className="auth__alert-error">
          {msgError}
        </div>)
      }

        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input" 
          value={password} 
          onChange={handleInputChange}     
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          Login
        </button>

        {/*div containing copied button we will use to authorize an user with their google account */}
        <div className="auth__social-networks">
          <p>Login with social networks</p>

          <div 
              className="google-btn"
              onClick={handleGoogleLogin}
          >
              <div className="google-icon-wrapper">
                  <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
              </div>
              <p className="btn-text">
                  <b>Sign in with google</b>
              </p>
          </div>

        </div>

        <Link 
          className="link"
          to="/auth/register"
        >
          Create new account
        </Link>
      </form>
  </>
  )
};
