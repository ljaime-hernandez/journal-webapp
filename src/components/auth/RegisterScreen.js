import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { setError, removeError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const {msgError} = useSelector( state => state.ui );

  const [values, handleInputChange] = useForm({
    name: 'Miguel',
    email: 'email@email.com',
    password: '123456',
    password2: '123456',
  })

  const {name, email, password, password2} = values;

  const handleRegister = (e) => {
    e.preventDefault();

    if(isFormValid()){
      dispatch(startRegisterWithEmailPasswordName(email, password, name))
    }
  }

  const isFormValid = () => {

    if(name.trim().length === 0){
      dispatch(setError("The name input is incorrect"))
      return false;
    } else if ( !validator.isEmail(email)){
      dispatch(setError("The email input is incorrect"))
      return false;
    } else if( password !== password2 || password2.length < 5){
      dispatch(setError("Password should be longer than 4 characters"))
      return false;
    }

    dispatch(removeError())
    return true;
  }

  return (
    <>
    <h3 className="auth__title">Register</h3>

    <form onSubmit={handleRegister}>

      {
        msgError &&
        (<div className="auth__alert-error">
          {msgError}
        </div>)
      }

    <input
        type="text"
        placeholder="Name"
        name="name"
        className="auth__input"
        autoComplete="off"
        value={name}
        onChange={handleInputChange}
        
      />

      <input
        type="text"
        placeholder="Email"
        name="email"
        className="auth__input"
        autoComplete="off"
        value={email}
        onChange={handleInputChange}
      />

      <input
        type="password"
        placeholder="Password"
        name="password"
        className="auth__input"
        value={password}
        onChange={handleInputChange}       
      />

<input
        type="password"
        placeholder="Confirm your Password"
        name="password2"
        className="auth__input"      
        value={password2}
        onChange={handleInputChange} 
      />

      <button
        type="submit"
        className="btn btn-primary btn-block mb-5"
      >
        Register
      </button>

      <Link 
        className="link"
        to="/auth/login"
      >
        Already registered?
      </Link>
    </form>
</>
  )
}
