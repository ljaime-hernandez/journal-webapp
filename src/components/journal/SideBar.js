import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries';

export const SideBar = () => {

    // useSelector retrieves the auth object from the store to render it in the
    // sidebar
    const {name} = useSelector( state => state.auth );
    // dispatch will run all our functions to the store using the respective
    // reducer
    const dispatch = useDispatch();
    // the dispatch will use an action which will delete the users
    // information on the reducer
    const handleLogout = () => {
        dispatch(startLogout());
    }
    // the new note will connect to the firebase database, then to the
    // user collection and then will create a document with the respective
    // journal information
    const handleAddNew = () => {
        dispatch(startNewNote());
    }

  return (
    <aside className="journal__sidebar">
        
        <div className="journal__sidebar-navbar">
            <h3 className="mt-5">
                <i className="far fa-moon"/>
                <span>{name}</span>
            </h3>

            <button
                className="btn" 
                onClick={handleLogout}   
            >
                Logout
            </button>
        </div>

        <div
            className="journal__new-entry"    
            onClick={handleAddNew}
        >
            <i className="far fa-calendar-plus fa-5x"></i>
            <p className="mt-5"></p>
            <p>New Entry</p>
            
        </div>

        <JournalEntries/>
    </aside>
  )
};
