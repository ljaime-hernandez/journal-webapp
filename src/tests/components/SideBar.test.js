import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { SideBar } from "../../components/journal/SideBar";
import { startLogout } from "../../actions/auth";
import { startNewNote } from "../../actions/notes";

jest.mock('../../actions/auth', () => ({
    startLogout: jest.fn()
}))
jest.mock('../../actions/notes', () => ({
    startNewNote: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
    auth: {
        uid:"HTXx027Mr1U94zHwXIsXNpT0M843",
        displayName:"Miguel"
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: {
            id: 'ABDCE',
            notes: []
        }
    }
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

describe('tests on SideBar component', () => {

    beforeEach(()=> {
        store = mockStore(initialState);
        jest.clearAllMocks();
    });
    
    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <SideBar/>
            </MemoryRouter>
        </Provider>
        );
    test('should render properly', () => {
      
        expect(wrapper).toMatchSnapshot();
    });

    test('should call logout', () => {
      
        wrapper.find('button').simulate('click');
        expect(startLogout).toHaveBeenCalled();
    });
    
    test('should call startNewNote', () => {
      
        wrapper.find('.journal__new-entry').simulate('click');
        expect(startNewNote).toHaveBeenCalled();
    });
    
})