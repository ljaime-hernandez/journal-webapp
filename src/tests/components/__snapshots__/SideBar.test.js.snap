// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on SideBar component should render properly 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="fas fa-portrait"
      >
          
      </i>
      <span>
          
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-alt fa-4x"
    />
    <p
      className="mt-5"
    />
    <p>
      New Entry
    </p>
  </div>
  <p
    className="journal__new-entry-p"
  >
    (scroll down to see all entries)
  </p>
  <div
    className="journal__entries animate__animated animate__fadeInLeft"
  />
</aside>
`;
